File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> lastIdx = new HashMap<>();
        int left = 0, maxLen = 0;

        for (int right = 0; right < s.length(); right++) {
            char c = s.charAt(right);

            if (lastIdx.containsKey(c)) {
                left = Math.max(left, lastIdx.get(c) + 1);
            }

            lastIdx.put(c, right);
            maxLen = Math.max(maxLen, right - left + 1);
        }

        return maxLen;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        lastSeen = dict()
        left = 0
        answer = 0

        for right in range(len(s)):
            c = s[right]

            if c in lastSeen and lastSeen[c] >= left:
                left = lastSeen[c] + 1
            
            lastSeen[c] = right
            answer = max(answer, right - left + 1)

        return answer
        